File: pkg/dual_endpoint/4client_spend_tx_update.go
package chain_utils

import (
	"fmt"

	ec "github.com/bsv-blockchain/go-sdk/primitives/ec"
	"github.com/bsv-blockchain/go-sdk/transaction"

	// primitives "github.com/bsv-blockchain/go-sdk/primitives/ec"
	multisig "github.com/spycat55/KeymasterMultisigPool/pkg/libs"

	tx "github.com/bsv-blockchain/go-sdk/transaction"
	sighash "github.com/bsv-blockchain/go-sdk/transaction/sighash"
)

// 最终 locaktime
const FINAL_LOCKTIME uint32 = 0xffffffff

// 合成两个签名
func LoadTx(
	txHex string,
	locktime *uint32,
	sequenceNumber uint32,
	serverAmount uint64,
	serverPublicKey *ec.PublicKey,
	clientPublicKey *ec.PublicKey,
	targetAmount uint64,
	// serverSignByte *[]byte,
	// clientSignByte *[]byte,
) (*transaction.Transaction, error) {
	// 恢复 bTx
	bTx, err := transaction.NewTransactionFromHex(txHex)
	if err != nil {
		return nil, err
	}

	if locktime != nil {
		bTx.LockTime = *locktime
	}

	// 创建优先级脚本
	priorityScript, err := multisig.Lock([]*ec.PublicKey{serverPublicKey, clientPublicKey}, 2)
	if err != nil {
		return nil, fmt.Errorf("创建优先级脚本失败: %v", err)
	}

	bTx.Inputs[0].SetSourceTxOutput(
		&tx.TransactionOutput{
			Satoshis:      targetAmount,
			LockingScript: priorityScript,
		},
	)

	// signs := [][]byte{*serverSignByte, *clientSignByte}
	// unScript, err := multisig.BuildSignScript(&signs)
	// if err != nil {
	// 	return nil, fmt.Errorf("BuildSignScript error: %v", err)
	// }

	// 更新输入
	// bTx.Inputs[0].UnlockingScript = unScript
	bTx.Inputs[0].SequenceNumber = sequenceNumber

	// 更新输出金额
	allAmount := bTx.Outputs[0].Satoshis + bTx.Outputs[1].Satoshis
	bTx.Outputs[0].Satoshis = serverAmount
	bTx.Outputs[1].Satoshis = allAmount - serverAmount

	// fmt.Printf("bTx 2: %s\n", bTx.Hex())

	return bTx, nil
}

// 双端费用池，分配资金, 客户端签名
// client -> server 修改金额和版本号
func ClientDualFeePoolSpendTXUpdateSign(
	tx *tx.Transaction,
	clientPrivateKey *ec.PrivateKey,
	serverPublicKey *ec.PublicKey,
) (*[]byte, error) {
	// if locktime != nil {
	// 	tx.LockTime = *locktime
	// }

	sigHash := sighash.Flag(sighash.ForkID | sighash.All)
	aMultisigUnlockingScriptTemplate, err := multisig.Unlock([]*ec.PrivateKey{}, []*ec.PublicKey{serverPublicKey, clientPrivateKey.PubKey()}, 2, &sigHash)
	if err != nil {
		return nil, fmt.Errorf("failed to create unlocking script template: %w", err)
	}

	// 重新签名所有输入
	clientSignByte, err := aMultisigUnlockingScriptTemplate.SignOne(tx, 0, clientPrivateKey)
	if err != nil {
		return nil, fmt.Errorf("c 重新签名输入 %d 失败: %v", 1, err)
	}

	return clientSignByte, nil
}

File: src/dual_endpoint/4client_spend_tx_update.ts
import { PrivateKey, PublicKey } from '@bsv/sdk/primitives';
import Transaction from '@bsv/sdk/transaction/Transaction';
import TransactionSignature from '@bsv/sdk/primitives/TransactionSignature';
import Script from '@bsv/sdk/script/Script';
import LockingScript from '@bsv/sdk/script/LockingScript';
import { hash256 } from '@bsv/sdk/primitives/Hash';
import * as ECDSA from '@bsv/sdk/primitives/ECDSA';
import BigNumber from '@bsv/sdk/primitives/BigNumber';
import { createDualMultisigScript } from './1base_tx';

export const FINAL_LOCKTIME = 0xffffffff;

/**
 * 载入 B-Tx（hex）并修改金额 / sequence / locktime 等信息
 * 对应 Go: LoadTx
 */
export function loadTx(
  txHex: string,
  locktime: number | undefined,
  sequenceNumber: number,
  serverAmount: number,
  serverPublicKey: PublicKey,
  clientPublicKey: PublicKey,
  targetAmount: number,
): Transaction {
  const tx = Transaction.fromHex(txHex);

  if (locktime !== undefined) {
    tx.lockTime = locktime;
  }

  // 设置源输出信息
  const priorityScript: Script = createDualMultisigScript([serverPublicKey, clientPublicKey]);
  const priorityLocking = new LockingScript();
  priorityLocking.chunks = priorityScript.chunks;

  if (!tx.inputs[0].sourceTransaction) {
    tx.inputs[0].sourceTransaction = new Transaction();
    tx.inputs[0].sourceTransaction.outputs = [];
  }

  tx.inputs[0].sourceTransaction.outputs[0] = {
    satoshis: targetAmount,
    lockingScript: priorityLocking,
  };

  tx.inputs[0].sequence = sequenceNumber;

  // 更新输出金额（index 0 server, index 1 client）
  const total = (tx.outputs[0].satoshis || 0) + (tx.outputs[1].satoshis || 0);
  tx.outputs[0].satoshis = serverAmount;
  tx.outputs[1].satoshis = total - serverAmount;

  return tx;
}

/**
 * 客户端在更新后的 B-Tx 上重新签名
 * 对应 Go: ClientDualFeePoolSpendTXUpdateSign
 */
export function clientDualFeePoolSpendTXUpdateSign(
  tx: Transaction,
  clientPrivateKey: PrivateKey,
  serverPublicKey: PublicKey,
): number[] {
  const clientPublicKey = clientPrivateKey.toPublicKey();

  const priorityScript = createDualMultisigScript([serverPublicKey, clientPublicKey]);

  const sighashData = TransactionSignature.format({
    sourceTXID: tx.inputs[0].sourceTXID || '',
    sourceOutputIndex: tx.inputs[0].sourceOutputIndex,
    sourceSatoshis: tx.inputs[0].sourceTransaction?.outputs[0].satoshis || 0,
    transactionVersion: tx.version,
    otherInputs: [],
    outputs: tx.outputs,
    inputIndex: 0,
    subscript: priorityScript,
    inputSequence: tx.inputs[0].sequence || 1,
    lockTime: tx.lockTime,
    scope: TransactionSignature.SIGHASH_ALL | TransactionSignature.SIGHASH_FORKID,
  });

  // 客户端确定性签名（RFC6979）
  const msgHash = hash256(sighashData);
  const signature = ECDSA.sign(new BigNumber(msgHash, 16), clientPrivateKey, true);
  const signatureDER = signature.toDER() as number[];

  return [
    ...signatureDER,
    TransactionSignature.SIGHASH_ALL | TransactionSignature.SIGHASH_FORKID,
  ];
}
